### PR TITLE
#264 :heart: #257

### PR DESCRIPTION
**Pull request by armanbilge**
Marries #264 and #257. Only remaining thing to consider is adding `map` and `contramap` to `LookupKey` and `InsertKey`.

cc @tpolecat

### CODE REVIEW

File: core/src/main/scala/org/typelevel/vault/InvariantMapping.scala [hunk 1]
```diff
@@ -21,15 +21,17 @@
...
   type I
   def in: A => I
   def out: I => A
   def imap[B](f: A => B)(g: B => A): InvariantMapping[B] =
     new InvariantMapping[B] {
       type I = outer.I
-      val in = g andThen outer.in
-      val out = outer.out andThen f
+      val in = AndThen(g) andThen outer.in
+      val out = AndThen(outer.out) andThen f
```
Comment (tpolecat) [inline core/src/main/scala/org/typelevel/vault/InvariantMapping.scala:34]: Neato, I didn't know about that one.

File: core/src/main/scala/org/typelevel/vault/Key.scala [hunk 2]
```diff
@@ -21,17 +21,84 @@
...
   override def hashCode(): Int = unique.hashCode()
+
+}
+
+sealed trait InsertKey[-A] { outer =>
+  private[vault] def unique: Unique.Token
+  private[vault] type I
+  private[vault] def in: A => I
+
+  def contramap[B](f: B => A): InsertKey[B] =
```
Comment (armanbilge) [inline core/src/main/scala/org/typelevel/vault/Key.scala:67]: Note that `contramap` and `map` will now be exposed on `Key` itself. Note that it's safe, because you'll end up with an `InsertKey`/`LookupKey` after doing this, but it could be confusing.

**Comment by armanbilge** [comment]
Cool, that wasn't bad at all. I think this is good to go.

**Comment by tpolecat** [comment]
Are we waiting on https://github.com/typelevel/cats/pull/3987 here? If not are there any other blockers?

**Comment by armanbilge** [comment]
We're not. Actually, I didn't (want to) use any coyoneda here since that would mean adding a new dependency to cats-free, unless that's okay. So, no blockers.

File: .gitignore [hunk 3]
```diff
@@ -11,3 +11,6 @@ tags
 .vscode
 .bsp
 .jekyll-cache/
+
+# metals
+metals.sbt
```
Comment (rossabaker) [inline .gitignore:16]: 💡 I can't remember who I've told, but ~/.config/git/ignore takes care of this on all your repos.

File: core/src/main/scala/org/typelevel/vault/Locker.scala [hunk 4]
```diff
@@ -39,15 +39,31 @@ final class Locker private (private val unique: Unique.Token, private val a: Any
...
   /**
    * Put a single value into a Locker
    */
-  def lock[A](k: Key[A], a: A): Locker = new Locker(k.unique, a.asInstanceOf[Any])
+  def lock[A](k: InsertKey[A], a: A): Locker = new Locker(k.unique, k.in(a))
+
+  /**
+   * Put a single value into a Locker
+   */
+  def lock[A](k: Key[A], a: A): Locker = lock(k: InsertKey[A], a)
```
Comment (rossabaker) [inline core/src/main/scala/org/typelevel/vault/Locker.scala:66]: Overloads make me queasy and overloads where one is a subtype of another make me queasier still.  I guess this is necessary for bincompat?  Is there a different name we could give the `InsertKey` version?

Comment (rossabaker) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: Or could we have just made this `private[vault]`?

Comment (rossabaker) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: We have the same basic issue with `unlock`.

File: core/src/main/scala/org/typelevel/vault/Vault.scala [hunk 5]
```diff
@@ -38,18 +38,38 @@ final class Vault private (private val m: Map[Unique.Token, Locker]) {
...
 
+  /**
+   * Delete a key from the vault
+   */
+  def delete[A](k: InsertKey[A]): Vault = Vault.delete(k, this)
+
+  /**
+   * Delete a key from the vault
+   */
+  def delete[A](k: LookupKey[A]): Vault = Vault.delete(k, this)
```
Comment (rossabaker) [inline core/src/main/scala/org/typelevel/vault/Vault.scala:71]: Uh, what happens if we call this with `Key`?  Does it pick one from the linearization of `Key`?  It doesn't really matter, I guess?

**Review by rossabaker — COMMENTED** [review]
I think this is good.  Some of the overloads are making me fidgety, but I can be convinced.

**Comment by armanbilge** [comment]
@rossabaker the spec on overload resolution is here: https://www.scala-lang.org/files/archive/spec/2.13/06-expressions.html#overloading-resolution

**tl;dr**: overloaded methods are resolved to the most specific available for the given type. This is why if there are overloads for _both_ `InsertKey` and `LookupKey` we must also provide an overload for `Key` to resolve the ambiguity.

File: core/src/main/scala/org/typelevel/vault/Locker.scala (hunk 4 above)
Comment (armanbilge) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: This is needed for bincompat, and no `private[vault]` does not work because it disappears the static forwarders. I ran into this same issue in https://github.com/typelevel/fs2/pull/2466.

OTOH, there is precedence for not caring about static forwarders in https://github.com/http4s/http4s/pull/5071#issuecomment-898705809.

File: core/src/main/scala/org/typelevel/vault/Vault.scala (hunk 5 above)
Comment (armanbilge) [reply core/src/main/scala/org/typelevel/vault/Vault.scala:71]: If we call this with `Key` it uses the existing overload for `Key` and ignores both of these.

Comment (rossabaker) [reply core/src/main/scala/org/typelevel/vault/Vault.scala:71]: Oh, I overlooked that there's still a `delete(Key[A])`.

File: core/src/main/scala/org/typelevel/vault/Locker.scala (hunk 4 above)
Comment (rossabaker) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: I'm not sure that was a good precedent.  And it seems we're stuck with the overload on `delete` anyway.  I find it a little confusing, but I don't see a good way out.

**Review by rossabaker — APPROVED** [review]

File: core/src/main/scala/org/typelevel/vault/Locker.scala (hunk 4 above)
Comment (armanbilge) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: Maybe an unpopular opinion, one way out would be to deprecate the vault companion object based methods (I always found the duplication confusing, does anyone actually use them like this?) in favor of just calling these methods directly on the vault instance, and then just not add these new methods to the companion object.

Comment (rossabaker) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: I didn't realize they existed on the companions.  I think that's a good idea.  The only weird one then is the overloaded `Lock.lock`, right? And perhaps the new one could be called `apply`?

Comment (armanbilge) [reply core/src/main/scala/org/typelevel/vault/Locker.scala:66]: Cool, let me give that a try 👍

**Comment by armanbilge** [comment]
@rossabaker I just pushed a commit that:
1. Bumps base version to 3.1 (not strictly necessary since everything is source-compatible, but we are making a lot of changes?).
2. Deprecates the companion object methods in favour of equivalent instance methods.
3. Avoids the `delete` overrides by introducing one more abstraction `DeleteKey` that is the parent of all keys.

**Review by rossabaker — APPROVED** [review]
:+1: if green

**Comment by armanbilge** [comment]
Good to go.

**Comment by rossabaker** [comment]
I'll leave Chris and Rob a short window to object, but I'd like to release it this weekend and unblock Rob's PR for an http4s release train on Monday.

**Comment by rossabaker** [comment]
Daniel, would love your thoughts here, too, if you've got time.

I think it's good, but this is all a slight move away from Tried And True Direct Haskell Port, so I'm being extra careful with it.

**Comment by rossabaker** [comment]
I'm going to release this.  We have a long-blocked and important http4s PR on this.

**Review by danicheg — COMMENTED** [review]
Sorry for being late (vacation in full swing). I think that's great work! And personally see no evil in moving away from `direct Haskell port` 🙈

**Comment by rossabaker** [comment]
Enjoy your break!
